### PR TITLE
[DebugBundle] Fix dump/die doesn't work with server dumper

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | 4.1 
| Bug fix?      | yes
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | N/A   
| License       | MIT
| Doc PR        | N/A

This is an attempt to fully respect the promise made by the server dumper feature, i.e: `dump` should act exactly the same when the `ServerDumper` is wired (`debug.dump_destination: "tcp://%env(VAR_DUMPER_SERVER)%"`) but the dump server is not up (`server:dump` not running).

Currently, when using the full stack and debug bundle, the promise is broken, as someone using `dd` or `dump` + `die` would not see expected dumps in the output for web requests. This is explained by the wrapped dumper (_the one used by `ServerDumper` when failing at contacting the server_) being the `var_dumper.cli_dumper`, so it outputs on `php://stdout` instead (which is fine for cli usages).

**To reproduce**

- `composer create-project "symfony/website-skeleton:^4.1@dev" no-dd-output`
- Add a controller and `dd` inside
- Start the webserver and access the controller => get a blank page, dumps are written to the webserver output.

**What this solution is worth**

It works, but is far from ideal. It reproduces the `DumpDataCollector` behavior (code is entirely borrowed from here) in a dedicated internal dumper used as the fallback when we failed at contacting the server. Also, it misses the file & line nb usually added by the collector in the output (_or would also need duplicating this logic_). **=> Now added**
Ideally, the `ServerDumper`'s wrapped dumper should be `DumpDataCollector` but it's not easily achievable as those are already quite intricated/tangled with each other in the way they are wired together. I failed to find a proper solution so far. 
=> _See https://github.com/symfony/symfony/pull/27614#issuecomment-398023101 last paragraph for a strategy extracting, deprecating & removing some code from DumpDataCollector, limiting it to its responsibility of collecting dumps to be shown on toolbar/profiler, in favor of the `DebugDumper` ([attempt in progress](https://github.com/symfony/symfony/compare/master...ogizanagi:4.2/refacto_dump_collector))._

Reverting #26675 makes things work, but the fix is legit and it would only rely on undesired behavior while still using the cli dumper as the wrapped dumper (thus leading to duplicates at the end of the process when one is calling a controller, e.g when executing a test suite).

### CODE REVIEW

**Comment by lyrixx** [comment]
I did not read the code, but:
1. I confirm there is a bug
2. I confirm this patch fix the bug

File: src/Symfony/Bundle/DebugBundle/VarDumper/DelayedDumper.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        $this->charset = $charset;
+        $this->flags = $flags;
+        $this->displayOptions = $displayOptions;
+        $this->cliDumper = $cliDumper;
+    }
+
+    public function dump(Data $data)
+    {
+        // only delay web requests
+        if (!\in_array(PHP_SAPI, array('cli', 'phpdbg'), true)) {
```
Comment (stof) [inline src/Symfony/Bundle/DebugBundle/VarDumper/DelayedDumper.php]: I would use `if (\PHP_SAPI === 'cli' || \PHP_SAPI === 'phpdbg')`, which can be resolved by OPCache at compile time

File: src/Symfony/Bundle/DebugBundle/VarDumper/DelayedDumper.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        $response = $event->getResponse();
+        $request = $event->getRequest();
+
+        // In all conditions that remove the web debug toolbar, dumps must be written on the output.
```
Comment (stof) [inline src/Symfony/Bundle/DebugBundle/VarDumper/DelayedDumper.php]: why would this service check the debug toolbar being injected or no ? This class is not related to the data collector AFAICT

Comment (ogizanagi) [reply src/Symfony/Bundle/DebugBundle/VarDumper/DelayedDumper.php]: You're right it's not directly related to the data collector, but this whole patch aims to ensure the exact same behavior exists whether the server dumper is wired or not.
When not wired and the toolbar being injected, dumps are already rendered in the toolbar and considered not to be output twice by the collector.

**Comment by ogizanagi** [comment]
After some discussion, closing in favor of #27614 with some compromises to avoid BC breaks.
Also, I won't explore more what's started https://github.com/symfony/symfony/compare/master...ogizanagi:4.2/refacto_dump_collector for now. Anyway will keep a trace if this can be re-considered in the future.
